### PR TITLE
[LTR] Do not add fields extracted using a query to the FieldValueFeatureExtractor

### PR DESCRIPTION
**Pull request by afoucret**
The PR has been repurposed to prevent the usage of field value extraction through `input.field_names`:

As a consequence, trying to put the following model:

```
    "input": { "field_names": [ "quantity" ] },
    "inference_config": {
      "learning_to_rank": {
        "num_top_feature_importance_values": 0,
        "feature_extractors": [
            {
              "query_extractor": {
                "feature_name": "tite_bm25",
                "query": {  "match": { "title": "{{query}}" } }
               }
            }
        ]
     }
  }
```
will cause a validation error when creating the model (`cannot specify [input.field_names] for a model of type [learning_to_rank]`).

Additional considerations:

- We will reintroduce a way to do direct field extraction post-GA using an API that is more consistent and can deal with situation such as multi-valued fields:

```
"feature_extractors": [
   { "query_extractor": { "feature_name": "title_bm25", "query": { ... } } }, # Query feature extractor
   { "field_value_extractor": { "feature_name": "quantity", "field": "quantity" } }, # Field value extractor
]
```

We prefer to disable the current behavior using `input.field_names` that will not be backward compatible.

- Previously, `input.field_names` was a mandatory non-empty field. When exporting models through `eland`, it was populated with all features names. A patch will created into eland to keep it empty when using ES >= 8.15.

### CODE REVIEW

**Comment by benwtrent** [comment]
@afoucret why isn't the patch to disallow duplicates on input vs. extracted queries? That should be the initial patch.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ltr/LearningToRankRescorerContext.java
Comment (afoucret) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ltr/LearningToRankRescorerContext.java:81]: // Should probably be commented until we have the new field value as a standard extractor.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceConfig.java
Comment (ioanatia) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceConfig.java:85]: to me it does not make much sense to have this method in here - its implementation seems to highly depend of what happens in `TrainedModelConfig`. Sending in the `validationException` also seems like a code smell.
Furthermore I don't see why it's an instance method - it does not depend on any of the instance variables, it could even be static?

Instead of moving this code from `TrainedModelConfig` - I would have added a simple `boolean requiresInputFields` method here that returns true and override it in `LearningToRankConfig`.
Then we could just use this method in `TrainedModelConfig` to do the validation.
With this approach we also don't need to add the `getDefaultInput` method to `InferenceConfig`.

**Review by kderusso — COMMENTED** [review]
Agreed with @ioanatia 's comments

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceConfig.java
Comment (afoucret) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceConfig.java:85]: I can remember another PR with the exact opposite point of view from Ben in a similar case (see comment https://github.com/elastic/elasticsearch/pull/109071#discussion_r1618802705).

**Comment by afoucret** [comment]
@elasticmachine rerun elasticsearch-ci/part-4

**Comment by kderusso** [comment]
@elasticmachine rerun elasticsearch-ci/part-4

**Comment by kderusso** [comment]
@elasticmachine merge upstream

**Review by kderusso — APPROVED** [review]
Approving this after discussion with the team on followup work, and conditional on passing CI before merging.

**Pull request merged**
